// Copyright (c) YugaByte, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License
// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied.  See the License for the specific language governing permissions and limitations
// under the License.
//
syntax = "proto2";

package yb.dockv;

option java_package = "org.yb.dockv";

import "yb/common/value.proto";

message ColumnPackingPB {
  optional uint32 id = 1;
  optional uint64 num_varlen_columns_before = 2;
  optional uint64 offset_after_prev_varlen_column = 3;
  optional uint64 size = 4;
  optional bool nullable = 5;

  // Present only in recent versions that were added after V2 implementation.
  optional PersistentDataType data_type = 6;
}

message SchemaPackingPB {
  optional uint32 schema_version = 1;
  repeated ColumnPackingPB columns = 2;

  // Columns that present in schema but don't participate in packing.
  repeated uint32 skipped_column_ids = 3;
}

enum DocdbLockMode {
  DOCDB_LOCK_EXCLUSIVE = 1;
  DOCDB_LOCK_SHARE = 2;
}
